src/farcasterd/trade_state_machine.rs

```diff
@@ -270,7 +302,40 @@ impl TradeStateMachine {
             TradeStateMachine::TakeOffer(TakeOffer { peerd, .. }) => Some(peerd.clone()),
             TradeStateMachine::TakerCommit(TakerCommit { peerd, .. }) => Some(peerd.clone()),
             TradeStateMachine::SwapdLaunched(SwapdLaunched { peerd, .. }) => Some(peerd.clone()),
-            TradeStateMachine::SwapdRunning(SwapdRunning { peerd, .. }) => Some(peerd.clone()),
+            TradeStateMachine::SwapdRunning(SwapdRunning { peerd, .. }) => peerd.clone(),
+            _ => None,
+        }
+    }
+
+    pub fn awaiting_connect_from(&self) -> Option<ServiceId> {
+        match self {
+            TradeStateMachine::TakerConnect(taker_connect) => Some(taker_connect.peerd.clone()),
+            TradeStateMachine::RestoringSwapd(RestoringSwapd {
+                public_offer,
+                trade_role,
+                expect_connection,
+                ..
+            }) => {
+                if *trade_role == TradeRole::Taker && *expect_connection {
+                    Some(ServiceId::Peer(node_addr_from_public_offer(&public_offer)))
+                } else {
+                    None
+                }
+            }
+            TradeStateMachine::SwapdRunning(SwapdRunning {
+                peerd,
+                public_offer,
+                ..
+            }) => {
+                // Peerd is none, implies a connect event is possible. A connect
+                // is only possible to a listening peer, its address thus has to
+                // match the address of the offer.
+                if peerd.is_none() {
+                    Some(ServiceId::Peer(node_addr_from_public_offer(public_offer)))
```

Comment:
with a taker that successively takes offers from the maker once the last completes: if the last successful run was Bob maker, then Alice taker will most likely terminate connection before Bob has sweeped his xmr. Then Alice launches another swap.

In my understanding, a new peer connection will not be registered because `awaiting_connect_from(..)` will still match against the tsm of the last successful run here, since Bob's peerd is down, yet `node_addr_from_public_offer(..)` is the same for the old and the new swap?

Reply:
I still don't understand quite understand how this triggers a state where the `peerd.is_none()`, but this has to be the case, so I added an additional check to only trigger this when case when taking / connecting a swap.